"No expansion needed for develop stage", but "AssertionError: Stage must be expanded"

### Steps to reproduce

```console
$ docker run --rm -it spack/ubuntu-noble:0.22.0
git clone -b acts-podio-schema-version https://github.com/eic/eic-spack
spack repo add $HOME/eic-spack
git clone -b v31.2.0 https://github.com/acts-project/acts
spack env activate --create acts
spack add acts
spack develop -p $HOME/acts acts@=31.2.0
spack concretize --fresh --force
spack -d install acts
```

*Edit:* shorter reproducer without external repo:
```console
$ docker run --rm -it spack/ubuntu-noble:0.22.0
cat << EOF >> $(spack location -p acts)/package.py
    patch("https://github.com/acts-project/acts/pull/3132.patch?full_index=1",
          sha256="e0c97940abc2b4eab50834f76dcfcf2e651b7bf961ad3dd6e124a56d4d5e1779")
EOF
git clone -b v31.2.0 https://github.com/acts-project/acts
spack env activate --create acts
spack add acts
spack develop -p $HOME/acts acts@=31.2.0
spack concretize --fresh --force
spack -d install acts
```

*Edit:* even shorter reproducer on `zlib` in https://github.com/spack/spack/issues/44942#issuecomment-2196565607

### Error message

```console
==> [2024-06-27-23:25:51.921311] Adding env modifications for acts
==> [2024-06-27-23:25:51.921368] setup_package: collected all modifications from dependencies
==> [2024-06-27-23:25:51.921429] A dependency has updated CPATH, this may lead pkg-config to assume that the package is part of the system includes and omit it when invoked with '--cflags'.
==> [2024-06-27-23:25:51.921576] setup_package: loading compiler modules
==> [2024-06-27-23:25:51.922402] '/usr/bin/gcc' '-v' '/tmp/spack-implicit-link-infoagjj3e8c/main.c' '-o' '/tmp/spack-implicit-link-infoagjj3e8c/output'
==> [2024-06-27-23:25:51.953330] implicit link dirs: link line:  /usr/libexec/gcc/x86_64-linux-gnu/13/collect2 -plugin /usr/libexec/gcc/x86_64-linux-gnu/13/liblto_plugin.so -plugin-opt=/usr/libexec/gcc/x86_64-linux-gnu/13/lto-wrapper -plugin-opt=-fresolution=/tmp/ccRfqLMM.res -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s -plugin-opt=-pass-through=-lc -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lgcc_s --build-id --eh-frame-hdr -m elf_x86_64 --hash-style=gnu --as-needed -dynamic-linker /lib64/ld-linux-x86-64.so.2 -pie -z now -z relro -o /tmp/spack-implicit-link-infoagjj3e8c/output /usr/lib/gcc/x86_64-linux-gnu/13/../../../x86_64-linux-gnu/Scrt1.o /usr/lib/gcc/x86_64-linux-gnu/13/../../../x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/13/crtbeginS.o -L/usr/lib/gcc/x86_64-linux-gnu/13 -L/usr/lib/gcc/x86_64-linux-gnu/13/../../../x86_64-linux-gnu -L/usr/lib/gcc/x86_64-linux-gnu/13/../../../../lib -L/lib/x86_64-linux-gnu -L/lib/../lib -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib -L/usr/lib/gcc/x86_64-linux-gnu/13/../../.. /tmp/ccpNuuN7.o -lgcc --push-state --as-needed -lgcc_s --pop-state -lc -lgcc --push-state --as-needed -lgcc_s --pop-state /usr/lib/gcc/x86_64-linux-gnu/13/crtendS.o /usr/lib/gcc/x86_64-linux-gnu/13/../../../x86_64-linux-gnu/crtn.o
==> [2024-06-27-23:25:51.953433] implicit link dirs: result: /usr/lib/gcc/x86_64-linux-gnu/13, /usr/lib/x86_64-linux-gnu, /usr/lib, /lib/x86_64-linux-gnu, /lib
==> [2024-06-27-23:25:51.954965] No fetching needed for develop stage.
==> [2024-06-27-23:25:51.954997] No checksum needed for develop stage.
==> [2024-06-27-23:25:51.955015] Sources for Develop stages are not cached
==> [2024-06-27-23:25:51.955029] No expansion needed for develop stage.
==> [2024-06-27-23:25:51.960687] Error: AssertionError: Stage must be expanded before applying patches
Traceback (most recent call last):
  File "/opt/spack/lib/spack/spack/build_environment.py", line 1151, in _setup_pkg_and_run
    return_value = function(pkg, kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/spack/lib/spack/spack/installer.py", line 2482, in build_process
    return installer.run()
           ^^^^^^^^^^^^^^^
  File "/opt/spack/lib/spack/spack/installer.py", line 2331, in run
    self.pkg.do_patch()
  File "/opt/spack/lib/spack/spack/package_base.py", line 1584, in do_patch
    patch.apply(self.stage)
  File "/opt/spack/lib/spack/spack/patch.py", line 292, in apply
    assert self.stage.expanded, "Stage must be expanded before applying patches"
           ^^^^^^^^^^^^^^^^^^^
AssertionError: Stage must be expanded before applying patches
==> [2024-06-27-23:25:51.961048] Flagging acts-31.2.0-modnyks2svgulod4qm3ix5wm2wet53wl as failed: AssertionError: Stage must be expanded before applying patches
==> [2024-06-27-23:25:51.969087] Reading config from file /opt/spack/var/spack/environments/acts/spack.yaml
==> [2024-06-27-23:25:51.971685] Deactivated environment 'acts'
==> [2024-06-27-23:25:51.981786] Using environment 'acts'
==> [2024-06-27-23:25:51.987527] '/opt/spack/var/spack/environments/acts/.spack-env/view/bin/git' '-C' '/opt/spack' 'rev-parse' 'HEAD'
==> [2024-06-27-23:25:52.007254] View at /opt/spack/var/spack/environments/acts/.spack-env/view does not need regeneration.
==> [2024-06-27-23:25:52.007434] ChildError: AssertionError: Stage must be expanded before applying patches
```

### Information on your system

* **Spack:** 0.22.0 (5fe93fee1eec46a0750bd340198bffcb92ff9eec)
* **Python:** 3.12.3
* **Platform:** linux-ubuntu24.04-skylake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

## It is not about the derived package

After moving the patch into the builtin repo, and removing the local repo, and then
```console
spack edit -N builtin acts
rm /root/.spack/repos.yaml
spack repo list
git -C $SPACK_ROOT diff
spack clean -a && spack concretize --fresh --force && spack install acts
```
we still get the same error, with only a patch addition to the builtin (I was surprised that there were no patches at all in `acts`).

This reduces the problem to just applying the following patch to the spack repo:
```diff
diff --git a/var/spack/repos/builtin/packages/acts/package.py b/var/spack/repos/builtin/packages/acts/package.py
index 1d252361..711b3f8b 100644
--- a/var/spack/repos/builtin/packages/acts/package.py
+++ b/var/spack/repos/builtin/packages/acts/package.py
@@ -176,6 +176,12 @@ class Acts(CMakePackage, CudaPackage):
     version("0.08.1", commit="289bdcc320f0b3ff1d792e29e462ec2d3ea15df6")
     version("0.08.0", commit="99eedb38f305e3a1cd99d9b4473241b7cd641fa9")
 
+    patch(
+        "https://github.com/acts-project/acts/pull/3132.patch?full_index=1",
+        sha256="e0c97940abc2b4eab50834f76dcfcf2e651b7bf961ad3dd6e124a56d4d5e1779",
+        when="@30.3.0:34",
+    )
+
     # Variants that affect the core Acts library
     variant(
         "benchmarks", default=False, description="Build the performance benchmarks", when="@0.16:"
```

<details><summary>Collapsed output</summary>
<pre>
root@373bc4276723:~# spack repo list
==> 1 package repository.
builtin    /opt/spack/var/spack/repos/builtin
root@373bc4276723:~# git -C $SPACK_ROOT diff
error: cannot run less: No such file or directory
diff --git a/var/spack/repos/builtin/packages/acts/package.py b/var/spack/repos/builtin/packages/acts/package.py
index 1d252361..711b3f8b 100644
--- a/var/spack/repos/builtin/packages/acts/package.py
+++ b/var/spack/repos/builtin/packages/acts/package.py
@@ -176,6 +176,12 @@ class Acts(CMakePackage, CudaPackage):
     version("0.08.1", commit="289bdcc320f0b3ff1d792e29e462ec2d3ea15df6")
     version("0.08.0", commit="99eedb38f305e3a1cd99d9b4473241b7cd641fa9")
 
&plus;    patch(
&plus;        "https://github.com/acts-project/acts/pull/3132.patch?full_index=1",
&plus;        sha256="e0c97940abc2b4eab50834f76dcfcf2e651b7bf961ad3dd6e124a56d4d5e1779",
&plus;        when="@30.3.0:34",
&plus;    )
&plus;
     &#35; Variants that affect the core Acts library
     variant(
         "benchmarks", default=False, description="Build the performance benchmarks", when="@0.16:"
root@373bc4276723:~# spack clean -a && spack concretize --fresh --force && spack install acts
==> Removing all temporary build stages
==> Removing cached downloads
==> Removing install failure marks
==> Removing cached information on repositories
==> Removing python cache files
==> Concretized acts
 -   rxtpl2r  acts@31.2.0%gcc@13.2.0~alignment~analysis~autodiff~benchmarks~cuda~edm4hep~fatras~fatras_geant4~identification~integration_tests~ipo~json~legacy~mlpack~odd~onnx~podio~profilecpu~profilemem~svg~sycl~unit_tests build_system=cmake build_type=Release cxxstd=17 dev_path=/root/acts generator=make log_failure_threshold=MAX patches=e0c9794 arch=linux-ubuntu24.04-skylake
[+]  cfpekk6      ^boost@1.85.0%gcc@13.2.0~atomic~chrono~clanglibcpp~container~context~contract~coroutine~date_time~debug~exception~fiber+filesystem~graph~graph_parallel~icu~iostreams~json~locale~log~math~mpi+multithreaded~nowide~numpy~pic+program_options~python~random~regex~serialization+shared~signals~singlethreaded~stacktrace~system~taggedlayout+test~thread~timer~type_erasure~versionedlayout~wave build_system=generic cxxstd=11 patches=a440f96 visibility=hidden arch=linux-ubuntu24.04-skylake
[+]  mc3inlf      ^cmake@3.27.9%gcc@13.2.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu24.04-skylake
[+]  kopokuj          ^curl@8.7.1%gcc@13.2.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu24.04-skylake
[+]  rgc6hdu              ^nghttp2@1.57.0%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  3l27ogs                  ^diffutils@3.10%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  aoehnw6              ^openssl@3.3.0%gcc@13.2.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu24.04-skylake
[+]  we7nwls                  ^ca-certificates-mozilla@2023-05-30%gcc@13.2.0 build_system=generic arch=linux-ubuntu24.04-skylake
[+]  t5g7ekx              ^perl@5.38.0%gcc@13.2.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu24.04-skylake
[+]  3xqhocp                  ^berkeley-db@18.1.40%gcc@13.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu24.04-skylake
[+]  2imzyrz                  ^bzip2@1.0.8%gcc@13.2.0~debug~pic+shared build_system=generic arch=linux-ubuntu24.04-skylake
[+]  yzt2vrg                  ^gdbm@1.23%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  omocjqo                      ^readline@8.2%gcc@13.2.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu24.04-skylake
[+]  hhimszz              ^pkgconf@2.2.0%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  dq7bb2u          ^ncurses@6.5%gcc@13.2.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-ubuntu24.04-skylake
[+]  esnol3m          ^zlib-ng@2.1.6%gcc@13.2.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  npde75x      ^eigen@3.4.0%gcc@13.2.0~ipo build_system=cmake build_type=RelWithDebInfo generator=make arch=linux-ubuntu24.04-skylake
[+]  rctl67v      ^gcc-runtime@13.2.0%gcc@13.2.0 build_system=generic arch=linux-ubuntu24.04-skylake
[+]  f6kk7bx      ^git-lfs@3.3.0%gcc@13.2.0 build_system=makefile arch=linux-ubuntu24.04-skylake
[+]  ek22kup          ^git@2.42.0%gcc@13.2.0+man+nls+perl+subtree~svn~tcltk build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  lhhbn4v              ^autoconf@2.72%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  klyytuq              ^automake@1.16.5%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  skvapnm              ^expat@2.6.2%gcc@13.2.0+libbsd build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  ary7vfe                  ^libbsd@0.12.1%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  upeusui                      ^libmd@1.0.4%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  lywagmk              ^gettext@0.22.5%gcc@13.2.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  qrl7uoy                  ^libxml2@2.10.3%gcc@13.2.0+pic~python+shared build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  m6vfess                  ^tar@1.34%gcc@13.2.0 build_system=autotools zip=pigz arch=linux-ubuntu24.04-skylake
[+]  lq4sgxk                      ^pigz@2.8%gcc@13.2.0 build_system=makefile arch=linux-ubuntu24.04-skylake
[+]  yknhbab                      ^zstd@1.5.6%gcc@13.2.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu24.04-skylake
[+]  7jq3sxi                  ^xz@5.4.6%gcc@13.2.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu24.04-skylake
[+]  wf6fx5e              ^libidn2@2.3.7%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  z6jn3uf                  ^libunistring@1.2%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  oa4hp2z              ^libtool@2.4.7%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  vs2yutl                  ^findutils@4.9.0%gcc@13.2.0 build_system=autotools patches=440b954 arch=linux-ubuntu24.04-skylake
[+]  2w5airz              ^m4@1.4.19%gcc@13.2.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu24.04-skylake
[+]  eaup2vu                  ^libsigsegv@2.14%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  pxk3pav              ^openssh@9.7p1%gcc@13.2.0+gssapi build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  xbpfe7k                  ^krb5@1.20.1%gcc@13.2.0+shared build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  g7kgizy                      ^bison@3.8.2%gcc@13.2.0~color build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  7hmomof                  ^libedit@3.1-20230828%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  qo7w2lc                  ^libxcrypt@4.4.35%gcc@13.2.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu24.04-skylake
[+]  ptu2bp7              ^pcre2@10.43%gcc@13.2.0~jit+multibyte build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  fmoa2nf          ^go@1.22.2%gcc@13.2.0 build_system=generic arch=linux-ubuntu24.04-skylake
[+]  fux2byr              ^go-bootstrap@1.20.6%gcc@13.2.0 build_system=generic arch=linux-ubuntu24.04-skylake
[e]  o2zy2ye      ^glibc@2.39%gcc@13.2.0 build_system=autotools arch=linux-ubuntu24.04-skylake
[+]  fvrdtlp      ^gmake@4.4.1%gcc@13.2.0~guile build_system=generic arch=linux-ubuntu24.04-skylake

[+] /usr (external glibc-2.39-o2zy2yejj6evkuqtja6724nnpwkgodkw)
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/gcc-runtime-13.2.0-rctl67v3i2zxipyyxl2rehavtti7r4rf
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/pcre2-10.43-ptu2bp77zsviumpod3j7w6saj4ze6ogc
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/ncurses-6.5-dq7bb2ujorcrijvctglfavdnsaogupcd
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/xz-5.4.6-7jq3sxitd7jjllooepz5kgnjjqdgzxys
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/boost-1.85.0-cfpekk6ut5y6ahlg3dsamwjr4ndygeha
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/zlib-ng-2.1.6-esnol3mpzuvtcnvt5o3a3ny2pzgikg6g
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/bzip2-1.0.8-2imzyrzikubryriqnkiu2tbr4srvn3x2
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/eigen-3.4.0-npde75xklgct5aabii7v6u5cjlyawnna
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/zstd-1.5.6-yknhbabpi5gil44i2lq5skz73cqqisrt
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/libunistring-1.2-z6jn3ufyd5kkczflmqxz6o7hppzbvb4l
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/nghttp2-1.57.0-rgc6hduxwck3fiqic2kxcpywwbgviktb
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/libmd-1.0.4-upeusuiciq3nliokvcetwmtvqwa5owac
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/libxcrypt-4.4.35-qo7w2lc4bsodrohiamzufn4yfern4gp6
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/berkeley-db-18.1.40-3xqhocp6kdbqgmfaievmoatirwfenxq4
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/gmake-4.4.1-fvrdtlptadancxuhttwzi3vkmunfu4qk
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/openssl-3.3.0-aoehnw67ee5ubwnetg3zanvsgdvlolgf
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/pigz-2.8-lq4sgxkh3cmfa2uhxreicxppemwb4eiq
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/libidn2-2.3.7-wf6fx5e4phkrnfjyjryiykcnkzrj3bvr
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/libbsd-0.12.1-ary7vfellvimhp6dtj2u2rl273i73oa6
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/libedit-3.1-20230828-7hmomof4f4555rd7j2zzjnkj5jijad4u
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/libxml2-2.10.3-qrl7uoyq7uuqjpk2btct4rmvgrjpngqo
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/readline-8.2-omocjqojzyvzjrzlbd7uahnwmngejejo
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/curl-8.7.1-kopokujkpy75t75m736lhfj63v25b47b
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/tar-1.34-m6vfessee75ufel2fzp2wxkihg6rtz2w
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/expat-2.6.2-skvapnmpjkkcb2v6lzavudiakvdt6zl3
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/gdbm-1.23-yzt2vrgqxfqpatojgtrqjuvrs3tb37jq
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/cmake-3.27.9-mc3inlfduqwlaeeei3p2pd75pcpbwuop
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/gettext-0.22.5-lywagmkgee46arevm6lwjcwxdjbtnem6
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/perl-5.38.0-t5g7ekx6n6vnxz5v5l2wspe2533tiydg
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/krb5-1.20.1-xbpfe7kpbdarn4q2eqrtcipfesfmbdii
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/openssh-9.7p1-pxk3pav6jbyhyl4vlnhetvhurprmmtsp
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/git-2.42.0-ek22kupxwmrlerm2doal5gddj5d5o5hs
[+] /opt/spack/opt/spack/linux-ubuntu24.04-skylake/gcc-13.2.0/git-lfs-3.3.0-f6kk7bxfrhfffnthcgjrpyqdzd45vwji
==> Installing acts-31.2.0-rxtpl2rzcerehety6n4yi7ad5vweamsw [35/35]
==> No binary for acts-31.2.0-rxtpl2rzcerehety6n4yi7ad5vweamsw found: installing from source
==> Error: AssertionError: Stage must be expanded before applying patches
</code></pre>
</details>

The issue is that normally a stage object is a composite, consisting of sources + resources + patches, but develop stages are not composite: so no resources / patches (for whatever reason) attached, meaning that they are not fetched and extracted upon install. `do_patch` loops over stage objects for patches anyhow and attempts to apply them, but they're not downloaded.

So `Error: AssertionError: Stage must be expanded before applying patches` makes sense: the patch itself was not staged.

I think we can just make it a composite stage with patches attached. IIUC, patches are allowed to fail for develop specs, and if they apply then repeated spack install should in principle skip patching because patching is stateful (it leaves some file to indicate that patches were applied).

Dunno about resources, preferably they're fetched too, cause I use `spack install x dev_path=/y` regularly to modify sources to get past build failures.

--- a/lib/spack/spack/package_base.py
+++ b/lib/spack/spack/package_base.py
@@ -1141,10 +1141,9 @@ class PackageBase(WindowsRPath, PackageViewMixin, RedistributionMixin, metaclass
             if not link_format:
                 link_format = "build-{arch}-{hash:7}"
             stage_link = self.spec.format_path(link_format)
-            return DevelopStage(compute_stage_name(self.spec), dev_path, stage_link)
-
-        # To fetch the current version
-        source_stage = self._make_root_stage(self.fetcher)
+            source_stage = DevelopStage(compute_stage_name(self.spec), dev_path, stage_link)
+        else:
+            source_stage = self._make_root_stage(self.fetcher)
 
         # all_stages is source + resources + patches
         all_stages = StageComposite()
@@ -1473,10 +1472,8 @@ class PackageBase(WindowsRPath, PackageViewMixin, RedistributionMixin, metaclass
             return
 
         checksum = spack.config.get("config:checksum")
-        fetch = self.stage.needs_fetching
         if (
             checksum
-            and fetch
             and (self.version not in self.versions)
             and (not isinstance(self.version, GitVersion))
         ):

--- a/lib/spack/spack/stage.py
+++ b/lib/spack/spack/stage.py
@@ -346,8 +346,6 @@ class Stage(LockableStagingDir):
     similar, and are intended to persist for only one run of spack.
     """
 
-    #: Most staging is managed by Spack. DevelopStage is one exception.
-    needs_fetching = True
     requires_patch_success = True
 
     def __init__(
@@ -772,8 +770,6 @@ class StageComposite(pattern.Composite):
                 "cache_mirror",
                 "steal_source",
                 "disable_mirrors",
-                "needs_fetching",
-                "requires_patch_success",
             ]
         )
 
@@ -813,6 +809,10 @@ class StageComposite(pattern.Composite):
         return self[0].archive_file
 
     @property
+    def requires_patch_success(self):
+        return self[0].requires_patch_success
+
+    @property
     def keep(self):
         return self[0].keep
 
@@ -823,7 +823,6 @@ class StageComposite(pattern.Composite):
 
 
 class DevelopStage(LockableStagingDir):
-    needs_fetching = False
     requires_patch_success = False
 
     def __init__(self, name, dev_path, reference_link):

--- a/lib/spack/spack/test/patch.py
+++ b/lib/spack/spack/test/patch.py
@@ -270,12 +270,9 @@ def trigger_bad_patch(pkg):
 def test_patch_failure_develop_spec_exits_gracefully(
     mock_packages, config, install_mockery, mock_fetch, tmpdir, mock_stage
 ):
-    """
-    ensure that a failing patch does not trigger exceptions
-    for develop specs
-    """
+    """ensure that a failing patch does not trigger exceptions for develop specs"""
 
-    spec = Spec("patch-a-dependency " "^libelf dev_path=%s" % str(tmpdir))
+    spec = Spec(f"patch-a-dependency ^libelf dev_path={tmpdir}")
     spec.concretize()
     libelf = spec["libelf"]
     assert "patches" in list(libelf.variants.keys())

